### PR TITLE
SSR: Cache State

### PR DESCRIPTION
**Pull request by ockham**
_tl;dr: Simplify server-side API data caching by leveraging Redux state serialization._

Previously, controllers of SSR'd sections were responsible for caching API data. But there's a better way: We're already sending Redux state to the client at server-render time, so why not also cache on the server side, and use that as the initial state when creating the Redux store for the next request (given that this is in logged-out mode, i.e. the cache key -- the route -- maps one-to-one to the state).

Previously, we'd be just [copying](https://github.com/Automattic/wp-calypso/blob/master/client/state/themes/actions.js#L174-L180) the `.then()` callback from the `requestThemes` action. So instead of doing this, I took a step back and came up with the present solution.

Furthermore, this PR removes `renderCacheKey` and just uses `context.path` for _markup_ cache. Previously, we were using concatenated path and API data (~sate) cache timestamp as key. Creating the markup cache key isn't the section controller's responsibility now. (If we still need the timestamp, we can add it back when caching in `server/render`.)

To test:
* Verify that Calypso still works.
* Enable `debug` for the server. I forgot how to do that 😊  Alternatively:
* Put a `console.log` [here](https://github.com/Automattic/wp-calypso/pull/10988/files#diff-b2d6784517d1151d7e75f0287d0c6478R134) to inspect `cachedServerState` (or `get( cachedServerState, 'themes.queries.wpcom.data')`), load `/themes` in your browser, watch your node console, reload `/themes` in your browser, compare.

TODO:
* Add logged-out check before using cached state as initial state in `server/pages`

### CODE REVIEW

**Comment by seear** [comment]
> Enable debug for the server

`export DEBUG="calypso:server-render"`

**Comment by seear** [comment]
Noting that in this PR, a render cache slot is taken up for every unique invalid path:

```
HEAD /version?1485513394331 200 1.323 ms - 20
  calypso:server-render cache access for key +19s /theme/rah
  calypso:server-render cache miss for key +0ms /theme/rah
  calypso:server-render Server render time (ms) +14ms 14
GET /theme/rah 200 569.208 ms - 15746
GET /calypso/style-debug.css?v=ecef94b6d3 304 1.744 ms - -
GET /calypso/vendor.development.js?v=147ad937b5 304 0.642 ms - -
GET /calypso/images/drake/drake-empty-results.svg 200 7.482 ms - 7847
HEAD /version?1485513413532 200 0.707 ms - 20
  calypso:server-render cache access for key +10s /theme/rahha
  calypso:server-render cache miss for key +0ms /theme/rahha
```

Whereas previously, an error string was used as the key:
```
  calypso:server-render cache access for key +9s theme not found
  calypso:server-render Server render time (ms) +0ms 0
GET /theme/rahha 304 551.650 ms - -
GET /calypso/style-debug.css?v=c97dd44bf8 304 1.158 ms - -
GET /calypso/vendor.development.js?v=147ad937b5 304 0.542 ms - -
GET /calypso/images/drake/drake-empty-results.svg 304 3.915 ms - -
HEAD /version?1485513578847 200 0.705 ms - 20
^[^[^[^[HEAD /version?1485513598850 200 1.080 ms - 20
  calypso:server-render cache access for key +25s theme not found
  calypso:server-render Server render time (ms) +0ms 0
GET /theme/blah 200 458.719 ms - 15657
```

**Comment by seear** [comment]
Previously, a timestamp in the key ensured that a render cache entry lasted no longer than one hour:

```
  calypso:server-render cache access for key +24s /theme/mood1485513775863
  calypso:server-render Server render time (ms) +0ms 0
GET /theme/mood 200 344.088 ms - 140909
```

With no timestamp in the key now, what causes a cache entry to go stale if theme details change?
```
  calypso:server-render cache access for key +3s /theme/mood
  calypso:server-render Server render time (ms) +0ms 0
GET /theme/mood 200 761.083 ms - 150701
```

**Comment by seear** [comment]
However this PR ends up, we should add some info around key requirements to the [docs](https://github.com/Automattic/wp-calypso/blob/master/docs/server-side-rendering.md).

**Comment by ockham** [comment]
@seear Great feedback, thanks!

> Noting that in this PR, a render cache slot is taken up for every unique invalid path: [...]

You're right -- this is where the one-to-one mapping of route to state/rendered markup breaks, and even worse so, since there can be infinitely many invalid routes.

I _think_ the optimal way to do this would be to have e.g. the theme sheet route fall thru to a `/theme/*` catch-all route-def, and have that render the 404. The cache key could then be a combination of `express`'s [`request.route.path`](http://stackoverflow.com/a/19460598) and `request.route.params` (or maybe just some serialization of the `request.route` object).

However, this will need a bit of tweaking. First of all, we can't currently fall thru an isoroute at all, since we always `render` implicitly after the last middleware, so we'd need to change that. (Luckily, I think that should be rather easy to do -- I think I made a PR for that at some point, tho that's probably outdated now.) Second, we should wait for the theme sheet to be componentized (#10748) so it'll be easier to move the not-found-error rendering to a different middleware and route.

So that's what I think would be the optimal solution -- curious to hear your thoughts.

I'll try and come up with an intermediary one for now 😁

**Comment by ockham** [comment]
(In other words, the 'optimal' solution would leverage the theme sheet's validation and routing to de-dupe the cache. I think that idea would scale to other sections, too -- they'd also just need to fall thru to their 404 routes, while the server renderer will continue to cache by `request.route.path` and `params` just fine.)

**Comment by seear** [comment]
> So that's what I think would be the optimal solution -- curious to hear your thoughts.

The most useful thing might be if I update the docs in a different PR with the cache requirements and how the current key scheme meets those, then you will be able to reason more quickly about alternative solutions.

**Comment by ockham** [comment]
Yeah, that'd be awesome :heart:

**Comment by seear** [comment]
Doc update: #11004

**Comment by ockham** [comment]
We should probably tackle error handling (#12863) before getting back to this.

File: server/pages/index.js [hunk 1]
```diff
@@ -38,14 +41,20 @@ const staticFiles = [
 
 let sections = sectionsModule.get();
 
+function getInitialServerState( serializedServerState ) {
+	// Bootstrapped state from a server-render
+	const serverState = reducer( serializedServerState, { type: DESERIALIZE } );
+	return pick( serverState, Object.keys( serializedServerState ) );
+}
```
Comment (ockham) [inline server/pages/index.js]: See https://github.com/Automattic/wp-calypso/pull/13339/files#r113051327

**Comment by ockham** [comment]
Closing in favor of #13339.

**Pull request closed without merging**